Add all releases and photon data to FENDL processing script

SamPUG opened the pull request:
This pull request contains changes related to issue #21 

I have included the following in the convert_fendl script:

- Photon data for FENDL-3.1a
- Photon data for FENDL-3.1d
- Neutron and Photon data for FENDL-3.0 
- Neutron and Photon data for FENDL-2.1

I also attempted to add FENDL-2.0 but faced problems because one of the energy laws required was not supported. This was discussed in #21 and an [issue](https://github.com/openmc-dev/openmc/issues/1487) was added to the main OpenMC repository 

I had some issues integrating FENDL-3.0 as K-39 ace file contained 'Inf' values within the data. OpenMC is unable to deal with this and cannot convert the file to HDF5. Although I looked into generating the file using the NJOY functionality built into OpenMC, it was decided that it was best to exclude K-39 (see #21 for more details). I have implemented a check and warning to test if the offending ace file is being processed. 

The FENDL-2.1 release is downloaded as individual ace files as it is not available in the same format as other releases. I thought it was best to create a download folder rather than download them in the current working directory.

Photo ENDF data for FENDL-2.1 is all within a single ENDF file. As such I added functionality which divides the file up into individual ENDF files before processing. I could not see a straight forward method of processing the file undivided with the incident photon class.

I've added a --cleanup and corresponding --no-cleanup argument which can be passed to remove the downloaded files. The --no-cleanup option is the default and its inclusion is slightly redundant but it follows the style of the other arguments. I chose to do --no-cleanup by default as this is the behavior of all the other convert scripts.

I have checked to make sure that the HDF5 files are generated for the added releases/particles and have checked the `cross_sections.xml` file contains the required libs.

Shimwell commented:
Just scanned over it and it looks good, I shall try out each of the new options soon. Thanks for making the PR

Shimwell commented on convert_fendl.py:
```diff
@@ -41,90 +42,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
                     "'earliest' for backwards compatibility or 'latest' for "
                     "performance")
-parser.add_argument('-r', '--release', choices=['3.1a', '3.1d'],
+parser.add_argument('-r', '--release', choices=['3.1a', '3.1d', '3.0', '2.1'],
```
rearrange this so they are in order of release newest to oldest
'3.1d', '3.1a', '3.0', '2.1'

Shimwell replied:
Could the description text on line 15-17 also be updated so it includes all the versions

Shimwell replied:
Thanks

Shimwell commented on convert_fendl.py:
```diff
@@ -41,90 +42,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
-                 as_browser=True, context=ssl._create_unverified_context())
+
+    for particle in args.particles:
+        if args.release == '2.1':
```
nice idea to have a download folder, I think we can roll this out to other scripts in the future. The number of compressed files we end up with in the data folder gets a bit excessive when downloading lots of libraries. Can put the .zip files into such a folder for all releases in the fendl script?

Shimwell replied:
this will help remove lots of the if args.release=2.1 cases

SamPUG replied:
Ok, currently the download path is named based on the release. At what level should the download folder be general? Would you like the download folder to be something like fendl-download instead which would imply that only FENDL data would go in there, or just call it something like downloads, so that other scripts could then use the same folder.

SamPUG replied:
In either case, it may still be worth having an extra folder for the 2.1 release as the ace files don't have any indication in the name that they are from FENDL. I think they clutter the directory and it is possible that they could clash with other libraries due to the generic naming.

Shimwell replied:
> Ok, currently the download path is named based on the release. At what level should the download folder be general? Would you like the download folder to be something like fendl-download instead which would imply that only FENDL data would go in there, or just call it something like downloads, so that other scripts could then use the same folder?

Based on release like you have is perfect for me

Shimwell commented on convert_fendl.py:
```diff
@@ -41,90 +42,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+        ]
+        for filename in sorted(neutron_files):
+            # Check for Inf values in K-39 ace file for FENDL-3.0
+            if args.release == '3.0' and filename.name == '19K_039.ace':
```
I was wondering if it is possible to move these odd cases into the release_details dictionary. I can't quite figure out how to do it elegantly but if you have any thoughts on how we can fully specify the release (including odd files and how to handle them) in the release_details then this would bring us one step closer to a generic nuclear data processing script

Shimwell replied:
No worries if this is too tricky, we do use this approach elsewhere so I am sure it is fine.

Shimwell commented on convert_fendl.py:
```diff
@@ -41,90 +42,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+    '2.1': {
+        'neutron':{
+            'base_url': 'https://www-nds.iaea.org/fendl21/fendl21mc/',
+            'files': ['H001mc.zip',  'H002mc.zip',  'H003mc.zip',  'He003mc.zip', 
```
Such a shame FENDL didn't release a single zip file for this release, but great to see the openmc scripts are flexible to accommodate such an input.

Shimwell commented on convert_fendl.py:
```diff
@@ -41,90 +42,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+        rmtree(download_path)
+
+    # Seperate photon ENDF file for 2.1 release
+    if args.release == '2.1' and 'photon' in args.particles:
```
oh wow FENDL 2.1 really bundled everything together like that, wonder if this is worth putting this logic into a function in case we come across this combine ENDF file approach again and keep the flow of the code easier to follow

SamPUG replied:
I could sort something out and make a PR to the main OpenMC repo. It would probably fit well in the `data.endf.py` module.

Shimwell replied:
Sounds good to me, what do you think @paulromano

SamPUG replied:
Just a note here - In theory, we could also use the `get_evaluations` function from the endf module and avoid splitting the file entirely.  Personally I prefer splitting it down so the processing is the same as the other convert scripts.

paulromano replied:
What about always using `get_evaluations`, even when we don't expect there to be more than one evaluation in the file? Seems like that would handle both cases.

Shimwell replied:
Less code in the convert_fendl scripts sounds great to me

Shimwell commented:
I have made now downloaded all the FENDL releases and it worked :-)

Shimwell commented:
I really like the use of a library-release-downloads folder and am keen to role that out across the codes in the future, it keeps things a little tidier

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +42,325 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+
+# Print the K-39 warning at the end so it doesn't get lost in the conversion messages
+if warn_k39:
+    print(ace_error_warning)
```
Use the [`warnings.warn`](https://docs.python.org/3/library/warnings.html#warnings.warn) function from the standard library

SamPUG replied:
Is it possible to make warnings raised this way appear at the end of the script? The reason I implemented it like this is so that this warning didn't get lost in all of the writing/converting output lines. 

I saw that this might be possible with `warnings.catch_warnings` but this would mean putting the whole script within the `catch_warnings` context.

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +42,325 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+                # Check for the error in case user has provided a fixed version.
+                if 'Inf' in open(filename, 'r').read():
+                    ace_error_warning = """
+                    WARNING: {} contains 'Inf' values within the XSS array which 
```
If you use `warnings.warn` below, no need to add "WARNING" here
```suggestion
                    {} contains 'Inf' values within the XSS array which 
```

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +42,325 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+                if 'Inf' in open(filename, 'r').read():
+                    ace_error_warning = """
+                    WARNING: {} contains 'Inf' values within the XSS array which 
+                    prevent conversion to a hdf5 file format. This is a known issue
```
```suggestion
                    prevent conversion to a HDF5 file format. This is a known issue
```

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +42,325 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+                    ace_error_warning = """
+                    WARNING: {} contains 'Inf' values within the XSS array which 
+                    prevent conversion to a hdf5 file format. This is a known issue
+                    in FENDL-3.0. {} has not been added to the cross section library 
```
```suggestion
                    in FENDL-3.0. {} has not been added to the cross section library.
```

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +42,325 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+        particle_details = release_details[args.release][particle]
+        for f in particle_details['files']:
+            download(urljoin(particle_details['base_url'], f),
+                    as_browser=True, context=ssl._create_unverified_context())
```
```suggestion
                     as_browser=True, context=ssl._create_unverified_context())
```

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +42,325 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+    for particle in args.particles:
+        # Create a directory to hold the downloads
+        particle_download_path = download_path / particle
+        particle_download_path.mkdir(parents = True, exist_ok=True) 
```
```suggestion
        particle_download_path.mkdir(parents=True, exist_ok=True) 
```

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +42,325 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+        
+        return created_files
+
+compressed_file_size, uncompressed_file_size = 0, 0
```
```suggestion
compressed_file_size = uncompressed_file_size = 0
```

SamPUG commented:
I've addressed changes from @paulromano and changed the handling of photon data from FENDL-2.1 to use the `openmc.data.endf.get_evaluations` function instead of splitting the ENDF file. 

I've also tried to include a general method for checking special cases. The idea is that handling functions are created for any specific special cases. These functions are specified in the `release_details` dictionary. Special cases are divided into the stage of the script where they can be applied. All special case functions should take a Path argument which gives the location of the file. 

Each special case function should return a dictionary with two keys: `'skip_file'` and `'err_msg'`. `'skip_file'` is a `bool` which is used to determine if this specific file should be ignored in processing and `'err_msg'` is a string which contains a warning. 

I think this method of dealing with special cases would handle most of the cases in other scripts so could help with generalising the data scripts. If you don't like it I can put back in the original way of handling the k-39 issue.

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +43,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+        err_msg = dedent(ace_error_warning)
+        skip_file = True
+
+    return {'skip_file': skip_file, 'err_msg': err_msg}
```
```suggestion
    if 'Inf' in open(file_path, 'r').read():
        ace_error_warning = """ 
        {} contains 'Inf' values within the XSS array 
        which prevent conversion to a HDF5 file format. This is a known issue
        in FENDL-3.0. {} has not been added to the cross section library. 
        """.format(file_path, file_path.name)
        err_msg = dedent(ace_error_warning)
        return {'skip_file': True, 'err_msg': err_msg}
    else:
        return {'skip_file': False}
```

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +43,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+            # Check if file requires special handling
+            if f in special_cases:
+                ret = special_cases[f](Path(f))
+                if len(ret['err_msg']) > 0:
```
```suggestion
                if 'err_msg' in ret:
```

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +43,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+            # Handling for special cases
+            if filename.name in special_cases:
+                ret = special_cases[filename.name](filename)
+                if len(ret['err_msg']) > 0:
```
```suggestion
                if 'err_msg' in ret:
```

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +43,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+            # Check if file requires special handling
+            if photo_path.name in special_cases:
+                ret = special_cases[photo_path.name](photo_path)
+                if len(ret['err_msg']) > 0:
```
```suggestion
                if 'err_msg' in ret:
```

paulromano commented on convert_fendl.py:
```diff
@@ -41,90 +43,308 @@ class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter,
+
+# Print any warnings
+for warning in output_warnings:
+        warnings.warn(warning, Warning)
```
```suggestion
    warnings.warn(warning)
```

paulromano commented:
@Shimwell do you want to take another look or should I go ahead and merge?

Shimwell commented:
This looks good to me, I shall changed the cwd parts once the new download function is available in _utilts.py. But that will be a separate PR
